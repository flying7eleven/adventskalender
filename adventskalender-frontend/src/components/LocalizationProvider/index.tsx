import React, { createContext, ReactNode } from 'react';

interface ResourceDefinitions {
    english: any;
    german: any;
}

interface Props {
    children: ReactNode[];
    resources: ResourceDefinitions;
}

interface LocalizationContextType {
    resources: ResourceDefinitions;
    translate: (key: string) => string;
}

export const LocalizationContext = createContext<LocalizationContextType>(null!);

export const LocalizationProvider = (props: Props) => {
    const deepFind = (obj: any, path: string) => {
        const paths = path.split('.');
        let current = obj;
        for (let i = 0; i < paths.length; ++i) {
            if (current[paths[i]] === undefined) {
                return undefined;
            } else {
                current = current[paths[i]];
            }
        }
        return current;
    };

    const getTranslationLanguage = () => {
        const navigatorLanguage = navigator.language.split('-');
        switch (navigatorLanguage[0].toLowerCase()) {
            case 'de':
                return props.resources.german;
            case 'en':
            default:
                return props.resources.english;
        }
    };

    const getTranslatedString = (key: string) => {
        // get the correct dictionary for the translation
        const translationLanguage = getTranslationLanguage();

        // get the translation for the key and handle if we did not find it
        const translatedValue = deepFind(translationLanguage, key);
        if (!translatedValue) {
            console.error(`Could not find localization for '${key}' (selected navigator language was ${navigator.language})`);
            return key;
        }

        // return the translated value
        return translatedValue;
    };

    const value = { resources: props.resources, translate: getTranslatedString };

    return <LocalizationContext.Provider value={value}>{props.children}</LocalizationContext.Provider>;
};
