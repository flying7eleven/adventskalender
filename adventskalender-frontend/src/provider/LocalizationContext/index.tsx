import { createContext } from 'react';

export interface ResourceDefinitions {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    english: any;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    german: any;
}

interface LocalizationContextType {
    resources: ResourceDefinitions;
    translate: (key: string) => string;
    translateWithPlaceholder: (key: string, placeholder: string) => string;
}

export const LocalizationContext = createContext<LocalizationContextType>(null!);
