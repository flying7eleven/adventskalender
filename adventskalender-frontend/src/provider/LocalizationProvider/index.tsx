import { ReactNode } from 'react';
import { LocalizationContext, ResourceDefinitions } from '../LocalizationContext';

interface Props {
    children: ReactNode | ReactNode[];
    resources: ResourceDefinitions;
}

export const LocalizationProvider = (props: Props) => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const deepFind = (obj: any, path: string) => {
        const paths = path.split('.');
        let current = obj;
        for (let i = 0; i < paths.length; ++i) {
            if (current[paths[i]] === undefined) {
                return undefined;
            } else {
                current = current[paths[i]];
            }
        }
        return current;
    };

    const getTranslationLanguage = () => {
        const navigatorLanguage = navigator.language.split('-');
        switch (navigatorLanguage[0].toLowerCase()) {
            case 'de':
                return props.resources.german;
            case 'en':
            default:
                return props.resources.english;
        }
    };

    const getTranslatedString = (key: string) => {
        // get the correct dictionary for the translation
        const translationLanguage = getTranslationLanguage();

        // get the translation for the key and handle if we did not find it
        const translatedValue = deepFind(translationLanguage, key);
        if (!translatedValue) {
            console.error(`Could not find localization for '${key}' (selected navigator language was ${navigator.language})`);
            return key;
        }

        // return the translated value
        return translatedValue;
    };

    const getTranslatedStringWithPlaceholder = (key: string, placeholder: string) => {
        return getTranslatedString(key).replace('%PLACEHOLDER%', placeholder);
    };

    const value = { resources: props.resources, translate: getTranslatedString, translateWithPlaceholder: getTranslatedStringWithPlaceholder };

    return <LocalizationContext.Provider value={value}>{props.children}</LocalizationContext.Provider>;
};
