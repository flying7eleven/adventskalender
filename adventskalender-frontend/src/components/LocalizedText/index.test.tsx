import { create } from 'react-test-renderer';
import { LocalizationProvider } from '../../provider/LocalizationProvider';
import { LocalizedText } from './index';

describe('LocalizedText component', () => {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    let languageGetter;
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    let originalError;

    beforeEach(() => {
        languageGetter = jest.spyOn(window.navigator, 'language', 'get');
        originalError = console.error;
        console.error = jest.fn();
    });

    afterEach(() => {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        console.error = originalError;
    });

    test('Matches the snapshot if known translation key and language is supplied', () => {
        const english = {
            existing: {
                key: 'Some translated Text',
            },
        };
        const german = {};
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        languageGetter.mockReturnValue('en');
        const translatedText = create(
            <LocalizationProvider resources={{ english, german }}>
                <div>
                    <LocalizedText translationKey={'existing.key'} />
                </div>
            </LocalizationProvider>
        );
        expect(translatedText.toJSON()).toMatchSnapshot();
    });

    test('Matches the snapshot if unknown translation key and valid language is supplied', () => {
        const english = {
            existing: {
                key: 'Some translated Text',
            },
        };
        const german = {};
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        languageGetter.mockReturnValue('en');
        const translatedText = create(
            <LocalizationProvider resources={{ english, german }}>
                <div>
                    <LocalizedText translationKey={'nonexisting.key'} />
                </div>
            </LocalizationProvider>
        );
        expect(translatedText.toJSON()).toMatchSnapshot();
    });
});
