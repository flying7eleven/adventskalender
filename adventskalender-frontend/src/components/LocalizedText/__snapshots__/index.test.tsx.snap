// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocalizedText component Matches the snapshot if known translation key and language is supplied 1`] = `
<div>
  Some translated Text
</div>
`;

exports[`LocalizedText component Matches the snapshot if unknown translation key and valid language is supplied 1`] = `
<div>
  nonexisting.key
</div>
`;
